Sources/StreamChat/Workers/Background/TokenHandler.swift

```diff
@@ -0,0 +1,225 @@
+    
+    private(set) var currentToken: Token?
+    
+    var connectionProvider: UserConnectionProvider
+    
+    // MARK: - Init & Deinit
+    
+    init(
+        connectionProvider: UserConnectionProvider,
+        retryStrategy: RetryStrategy,
+        maximumTokenRefreshAttempts: Int,
+        timerType: Timer.Type
+    ) {
+        self.connectionProvider = connectionProvider
+        self.retryStrategy = retryStrategy
+        self.maximumTokenRefreshAttempts = maximumTokenRefreshAttempts
+        self.retryTimerType = timerType
+    }
+    
+    deinit {
+        let error = ClientError.ClientHasBeenDeallocated()
+        cancelTokenWaiters(with: error)
+    }
+    
+    // MARK: - TokenHandler
+    
+    func set(token: Token, completion: ((Error?) -> Void)?) {
+        if let userId = connectionProvider.userId, token.userId != userId {
+            completion?(ClientError.InvalidToken("The token is for another user"))
+            return
+        }
+        
+        handleTokenResult(.success(token))
+        completion?(nil)
+    }
+    
+    func cancelTokenWaiters(with error: Error) {
+        handleTokenResult(.failure(error))
+    }
+    
+    func refreshToken(completion: @escaping TokenWaiter) {
+        log.debug("Token refresh is initiated.", subsystems: .tokenRefresh)
+        
+        let shouldTriggerRefresh = initiateRefreshIfNotRunning()
+        
+        _ = add(tokenWaiter: completion)
+        
+        guard shouldTriggerRefresh else {
+            log.debug("Token refresh process is already in progress.", subsystems: .tokenRefresh)
+            return
+        }
+        
+        let tokenBeforeRefresh = currentToken
+        
+        retry { [weak self] in
+            guard let self = self else { return }
+            
+            // When `set(token:)` get's invoked while the ongoing token refresh process,
+            // we should assign the given token and cancel the refresh flow. However, cancelling refresh flow
+            // is currently not possible because `tokenProvider` is not returning the cancellable (planned for v5).
+            //
+            // To avoid overriding the token assinged by `set(token:)`, we store the token before the refresh
+            // and compare it we with the current token when refresh is completed. If the tokens does not match,
+            // it means the token was manually assigned and results from `tokenProvider` should be discarded.
+            guard self.currentToken == tokenBeforeRefresh else {
+                let tokenAssingedWhileRefresh = """
+                The `currentToken` was assinged while the refresh process.
+                The token given by `tokenProvider` will be discarded.
+                """
+                log.info(tokenAssingedWhileRefresh, subsystems: .tokenRefresh)
+                return
+            }
+            
+            switch $0 {
+            case .success(let newToken) where newToken == self.currentToken:
```

Comment:
We could also check that the new token is for the current user (there'll be one more check related to `token.userId`)